bin/verify-binary-architecture.sh

```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env bash
+set -xeuo pipefail
+
+if [ $# -ne 2 ]; then
+  echo "Usage: $0 <target-triple> <binary-path>"
+  exit 1
+fi
+TARGET_TRIPLE="$1"
+BINARY_PATH="$2"
+
+# Parse target architecture
+case "${TARGET_TRIPLE}" in
+  "x86_64"*) TARGET_ARCHITECTURE="x86_64" ;;
+  "aarch64"*) TARGET_ARCHITECTURE="aarch64" ;;
+  *) echo "Unknown target architecture '${TARGET_TRIPLE}'"; exit 1 ;;
+esac
+
+# Parse binary architecture and map expected binary architecture for target OS
+case "${TARGET_TRIPLE}" in
+  *"darwin"*)
+    # Parse binary architecture
+    file_output=$(file -b "${BINARY_PATH}")
+    BINARY_ARCHITECTURE=$(echo "${file_output}" | grep -o "x86_64\|arm64" || echo "")
+    # Map expected binary architecture
+    case "${TARGET_ARCHITECTURE}" in
+      "x86_64") EXPECTED_BINARY_ARCHITECTURE="x86_64" ;;
+      "aarch64") EXPECTED_BINARY_ARCHITECTURE="arm64" ;;
+      *) echo "Unknown Darwin architecture: '${TARGET_ARCHITECTURE}'"; exit 1 ;;
+    esac
+    ;;
+  *"linux"*)
+    # Parse binary architecture
+    file_output=$(file -b "${BINARY_PATH}")
+    BINARY_ARCHITECTURE=$(echo "${file_output}" | grep -o "x86-64\|aarch64" | head -n1 || echo "")
+    # Map expected binary architecture
+    case "${TARGET_ARCHITECTURE}" in
+      "x86_64") EXPECTED_BINARY_ARCHITECTURE="x86-64" ;;
+      "aarch64") EXPECTED_BINARY_ARCHITECTURE="aarch64" ;;
+      *) echo "Unknown Linux architecture: '${TARGET_ARCHITECTURE}'"; exit 1 ;;
+    esac
+    ;;
+  *"windows"*)
+    # Parse binary architecture
+    pe_header_output=$(powershell -Command "
```

Comment:
Is `file` unavailable on the Windows runners, even though it has a posix shell installed?

Reply:
I am afraid not. Same is true for `which` vs `where`. It's been a pain getting this to work for Windows in the least verbose way.

That's the reason why I opted for creating separate workflows. So that I can test an even broader set of target architectures and OSes and make it work:
```yaml
target:
  - x86_64-unknown-linux-gnu
  - aarch64-unknown-linux-gnu
  - x86_64-unknown-linux-musl
  - aarch64-unknown-linux-musl
  - x86_64-apple-darwin
  - aarch64-apple-darwin
  - x86_64-pc-windows-msvc
  - aarch64-pc-windows-msvc
```

Reply:
Interesting. I added an isolated file output test workflow step on a separate testing branch:

```yaml
- name: Test file output
  shell: bash
  run: |
    set -xeuo pipefail

    BINARY_PATH="target/${{ matrix.target }}/release/gleam"
    if [[ "${{ matrix.target }}" == *"windows"* ]]; then
        BINARY_PATH="${BINARY_PATH}.exe"
    fi

    file_output=$(file "${BINARY_PATH}")
```

And ran it on the various target architectures and OSes:

```sh
target/x86_64-unknown-linux-gnu/release/gleam: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=7fa5e3ea50b409f61625cfaa5cf439a30de76936, for GNU/Linux 3.2.0, not stripped

target/aarch64-unknown-linux-gnu/release/gleam: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, for GNU/Linux 3.7.0, BuildID[sha1]=75394ba54846e1cd0654df8f49cf615896418733, not stripped

target/x86_64-unknown-linux-musl/release/gleam: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), static-pie linked, BuildID[sha1]=1892a4eef798936540c664af5c9f3f96d7253d6f, not stripped

target/aarch64-unknown-linux-musl/release/gleam: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), statically linked, not stripped

target/x86_64-apple-darwin/release/gleam: Mach-O 64-bit executable x86_64

target/aarch64-apple-darwin/release/gleam: Mach-O 64-bit executable arm64

target/x86_64-pc-windows-msvc/release/gleam.exe: PE32+ executable (console) x86-64, for MS Windows, 5 sections

target/aarch64-pc-windows-msvc/release/gleam.exe: PE32+ executable (console) Aarch64, for MS Windows, 11 sections

```

Seems like I was wrong – it works - sorry! Not sure why it hasn't previously. Let me fix that.

Reply:
Oups, I figured out why it appeared to not work when I did this initially... 🤦 

Anyways: refactored the bash script to use file on Windows too in commit https://github.com/gleam-lang/gleam/pull/3897/commits/425e8fbf14f041f7fe08dca671a3a2b959213eec.

Reply:
Classic CI programming problems! 2 parts, engineering, 1 part prayers